docs(demo): document Prometheus-client custom metric in ad service (#10033)
open-telemetry/opentelemetry.io: content/en/docs/demo/services/ad.md

  .build();
```

### Bridging non-OTel custom metrics (Prometheus client library)

The Ad service also exposes a small set of custom metrics using the
[Prometheus Java client library](https://github.com/prometheus/client_java)
rather than the OpenTelemetry SDK. These metrics are exposed on a separate HTTP
endpoint (`/metrics` on `AD_PROMETHEUS_PORT`, default `9465`) and scraped by the
OpenTelemetry Collector's
[`prometheus` receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver),
which forwards them into the same pipeline as the OTel SDK metrics:

```java
private static final Counter adsServedCounter =
    Counter.builder()
        .name("demo_ad_served_total")
        .help("Total number of ads served, labeled by category")
        .labelNames("category")
        .register();

HTTPServer prometheusServer =
    HTTPServer.builder().port(prometheusPort).buildAndStart();
```

> This is intentionally included to illustrate a **common pattern during OTel

Reviewer: I would consider marking this by https://github.com/orgs/community/discussions/16925

> OTel adoption**: organizations frequently already own significant Prometheus
> instrumentation -- in libraries, third-party exporters, or legacy services --
> and want to ingest those metrics into an OpenTelemetry-native pipeline without

> > rewriting everything up front. The Collector's `prometheus` receiver is the

Reviewer 2: ```suggestion
> rewriting everything upfront. The Collector's `prometheus` receiver is the
```

> > bridge that makes this possible.

Reviewer 3: The rendering of these notes turns text on the same line as the category into heading text, so in this case it looks a little weird. The suggestion should fix that.

<img width="777" height="244" alt="Image" src="https://github.com/user-attachments/assets/d72b6820-2b1d-46b7-ab8e-99b80a821f24" />

```suggestion
> [!NOTE] 
>
> This is intentionally included to illustrate a **common pattern during
> OTel adoption**: organizations frequently already own significant Prometheus
> instrumentation -- in libraries, third-party exporters, or legacy services --
> and want to ingest those metrics into an OpenTelemetry-native pipeline without
> rewriting everything upfront. The Collector's `prometheus` receiver is the
> bridge that makes this possible.
```

Author: thanks Tiffany, I tried with a single line break and the CI format checker was unhappy, that's why I put it on the same line. Let's try your suggestion.

The Collector configuration that wires this up:

```yaml
receivers:
  prometheus/ad:
    config:
      scrape_configs:
        - job_name: ad
          scrape_interval: 10s
          static_configs:
            - targets: ['ad:${env:AD_PROMETHEUS_PORT}']
```

> > [!TIP] **Recommendation**: this is a _transitional_ pattern. Prefer the

Reviewer 4: On the Prometheus SIG, we're working on making "scraping Prometheus and exporting OTLP" give you all the rich semantics you get by instrumenting with OTLP from the start. I understand this is not the case today, but we'll get there! 

This is your documentation, and if you'd like to word it like this, I have no objections! Just sharing that transitioning to an OTLP SDK just for the sake of "use OTel" might not be really relevant to the message here 😅

Author: Thanks for the review, Arthur.

We identified several challenges when mixing monitoring libraries. The ongoing OpenTelemetry ↔ Prometheus alignment work should reduce some of that complexity, but many of the challenges we discussed are likely to remain.

Would it make sense to expand this section to describe those trade-offs in more detail, so users can better understand the implications and make an informed decision for their own environment? (cc @Kielek @puckpuck )

Reviewer 4: Up to you :) 

If those problems you mentioned are unsolvable as you claim, no problem telling people to migrate.

From my standpoint it sounds weird this requirement unless for the obvious reasons, logs+traces+metrics in a single SDK, but I'm probably missing context here

Author: We will document the challenges and let users figure out what's best for them.
Amongst the challenges:
- Misalignment of `service.name` and `service.instance.id` which hopefully will be solved by @dashpole 's work:
   - https://github.com/open-telemetry/opentelemetry-specification/pull/4956
- Simultaneous usage of two semantically different metrics libraries and model (label vs attribute, different semconv...) with different APIs and ingestions pipelines and potentially different enrichment/transformations. It's particularly challenging to troubleshoot as metrics at rest in backends like `prometheus don't track the emitter (ie the scope name)
-  This heterogeneity would end up with code lacking consistency like a code block using simultaneously Prom Client APIs from some/older custom metrics and OTel APIs for span attributes and other/newer custom metrics

Reviewer 4: Thanks for the details, agreed with most of them! 

The scope problem is easily solvable by adding the "otel_scope_*" labels by the emitter or relabeling rules in Prometheus receiver. That said, also agree it's extra work that wouldn't be necessary if using OTel SDKs

SDK for new custom metrics, and migrate existing Prometheus-client metrics to it
-- incremental migration as you touch the surrounding code has proven successful

> in practice, but a focused refactor works too; what matters is that the

Reviewer: same here https://github.com/orgs/community/discussions/16925

> > works too; what matters is that the migration happens.

Reviewer 3: ```suggestion
> [!TIP] 
>
> **Recommendation**: this is a _transitional_ pattern. Prefer the
> OpenTelemetry SDK for new custom metrics, and migrate existing
> Prometheus-client metrics to it -- incremental migration as you touch the
> surrounding code has proven successful in practice, but a focused refactor
> works too; what matters is that the migration happens.
```
